### PR TITLE
feat(flag-decisions): Add support for sending flag decisions along with decision metadata.

### PR DESCRIPTION
**Pull request by yasirfolio3**
## Summary
* Send events for **ALL** decision types.
* Add `Metadata` field to EventBatch.Decisions to capture flag type, key and variation key.

## Test summary
* Added unit tests.

### CODE REVIEW

File: Sources/Data Model/Project.swift [hunk 1]
```diff
@@ -42,6 +42,7 @@ struct Project: Codable, Equatable {
     var typedAudiences: [Audience]?
     var featureFlags: [FeatureFlag]
     var botFiltering: Bool?
+    var sendFlagDecisions: Bool?
```
Comment (msohailhussain) [inline Sources/Data Model/Project.swift:45]: It's not nullable.

File: Sources/Implementation/Events/BatchEventBuilder.swift [hunk 2]
```diff
@@ -25,13 +25,29 @@ class BatchEventBuilder {
...
+                                      flagKey: String,
+                                      flagType: String) -> Data? {
+        
+        if (flagType == Constants.DecisionSource.rollout.rawValue || variation == nil) && !config.sendFlagDecisions() {
+            return nil
+        }
+
+        var variationId = ""
+        var variationKey = ""
+        if let tmpVariation = variation {
```
Comment (msohailhussain) [inline Sources/Implementation/Events/BatchEventBuilder.swift:40]: why using it? 
variation?.Id should be fine.

File: Sources/Optimizely/OptimizelyClient.swift [hunk 3]
```diff
@@ -372,38 +374,34 @@ open class OptimizelyClient: NSObject {
...
                                                           userId: userId,
-                                                          attributes: attributes ?? OptimizelyAttributes())
-        
-        guard let variation = pair?.variation else {
-            logger.i(.variationUnknown(userId, featureKey))
+                                                          attributes: attributes ?? OptimizelyAttributes()) else {
             sendDecisionNotification(decisionType: .feature,
                                      userId: userId,
                                      attributes: attributes,
+                                     source: Constants.DecisionSource.rollout.rawValue,
```
Comment (msohailhussain) [inline Sources/Optimizely/OptimizelyClient.swift:384]: need to check.

File: Sources/Optimizely/OptimizelyClient.swift [hunk 4]
```diff
@@ -771,12 +775,14 @@ extension OptimizelyClient {
             // send notification in sync mode (functionally same as async here since it's already in background thread),
             // but this will make testing simpler (timing control)
 
-            self.sendActivateNotification(experiment: experiment,
-                                          variation: variation,
-                                          userId: userId,
-                                          attributes: attributes,
-                                          event: event,
-                                          async: false)
+            if let tmpVariation = variation {
```
Comment (msohailhussain) [inline Sources/Optimizely/OptimizelyClient.swift:778]: Why need to add this condition?

**Review by msohailhussain — CHANGES_REQUESTED** [review]
please address.

File: Sources/Data Model/Project.swift (hunk 1 above)
Comment (yasirfolio3) [reply Sources/Data Model/Project.swift:45]: Since `Project` struct is marked codable, for every value that might not exist, we have to mark it as optional, otherwise decoding a datafile with no `sendFlagDecisions` into `Project` struct will cause a crash.

File: Sources/Implementation/Events/BatchEventBuilder.swift (hunk 2 above)
Comment (yasirfolio3) [reply Sources/Implementation/Events/BatchEventBuilder.swift:40]: Value of optional type 'String?' must be unwrapped to a value of type 'String' before assigning it into the struct. This change is there since variation can now be nil too.

File: Sources/Optimizely/OptimizelyClient.swift (hunk 4 above)
Comment (yasirfolio3) [reply Sources/Optimizely/OptimizelyClient.swift:778]: Variation was non-optional before so we didn't have to unwrap it, but since we have allowed impression event's to be sent with nil variation, this is necessary!

**Review by pawels-optimizely — DISMISSED** [review]
lgtm

File: Tests/OptimizelyTests-Common/DecisionServiceTests_Features.swift [hunk 5]
```diff
@@ -391,11 +401,11 @@ extension DecisionServiceTests_Features {
         self.config.project.rollouts[0].experiments[0].audienceIds = [kRolloutAudienceIdAge2]
         self.config.project.rollouts[0].experiments[0].trafficAllocation[0].endOfRange = 0
         self.config.project.rollouts[0].experiments[2].audienceIds = [kRolloutAudienceIdAge1]
-        let variation = self.decisionService.getVariationForFeatureRollout(config: config,
+        let pair = self.decisionService.getVariationForFeatureRollout(config: config,
                                                                            featureFlag: featureFlag,
                                                                            userId: kUserId,
                                                                            attributes: kAttributesRolloutAge2Match)
-        XCTAssertNil(variation)
+        XCTAssertNil(pair)
```
Comment (jaeopt) [inline Tests/OptimizelyTests-Common/DecisionServiceTests_Features.swift:408]: Can we add test cases for decision event sent for this nil pair cases?

File: Tests/OptimizelyTests-Common/BatchEventBuilderTests_Events.swift [hunk 6]
```diff
@@ -76,6 +76,12 @@ class BatchEventBuilderTests_Events: XCTestCase {
         XCTAssertEqual(decision["campaign_id"] as! String, expCampaignId)
         XCTAssertEqual(decision["experiment_id"] as! String, expExperimentId)
         
+        let metaData = decision["metadata"] as! Dictionary<String, Any>
+        XCTAssertEqual(metaData["rule_type"] as! String, "experiment")
+        XCTAssertEqual(metaData["rule_key"] as! String, "ab_running_exp_audience_combo_exact_foo_or_true__and__42_or_4_2")
+        XCTAssertEqual(metaData["flag_key"] as! String, "")
```
Comment (jaeopt) [inline Tests/OptimizelyTests-Common/BatchEventBuilderTests_Events.swift:82]: Does this test pass? I understand "flag_key" set to same as "rule_key" for activate.

**Review by jaeopt — CHANGES_REQUESTED** [review]
Overall looks good. See my comments for change requests.

File: Sources/Implementation/DefaultDecisionService.swift [hunk 7]
```diff
@@ -142,18 +142,17 @@ class DefaultDecisionService: OPTDecisionService {
...
         
         return nil
-        
     }
     
     func getVariationForFeatureExperiment(config: ProjectConfig,
                                           featureFlag: FeatureFlag,
                                           userId: String,
-                                          attributes: OptimizelyAttributes) -> (experiment: Experiment?, variation: Variation?)? {
+                                          attributes: OptimizelyAttributes) -> (experiment: Experiment, variation: Variation?, source: String)? {
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply Sources/Implementation/DefaultDecisionService.swift:155]: We can't since getVariation still returns `optional` Variation.

File: Sources/Implementation/Events/BatchEventBuilder.swift [hunk 8]
```diff
@@ -25,13 +25,29 @@ class BatchEventBuilder {
...
+        if (ruleType == Constants.DecisionSource.rollout.rawValue || variation == nil) && !config.sendFlagDecisions() {
+            return nil
+        }
         
-        let decision = Decision(variationID: varionation.id,
+        var variationId = ""
+        var variationKey = ""
+        if let tmpVariation = variation {
+            variationKey = tmpVariation.key
+            variationId = tmpVariation.id
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply Sources/Implementation/Events/BatchEventBuilder.swift:42]: yes we do now.

File: Sources/Optimizely/OptimizelyClient.swift [hunk 9]
```diff
@@ -385,8 +385,7 @@ open class OptimizelyClient: NSObject {
                                      featureEnabled: false)
             return false
```
Comment (jaeopt) [inline Sources/Optimizely/OptimizelyClient.swift:386]: We still need to send an impression event with a dummy metadata even when pair is null. These lines should be refactored to support it.

Comment (jaeopt) [reply Sources/Optimizely/OptimizelyClient.swift:386]: We also need to change the impression-event from activate call. When the experimentKey is for feature-test (not ab-test) we should send a valid flag-key (instead of copying experiment-key).

**Review by jaeopt — CHANGES_REQUESTED** [review]
Refactor for non-optional variation look great!
A couple of changes requested as the spec was clarified in the meeting today.
- it looks like a dummy event still not sent out when a variation is decided to nil (we also need a test case for this)
- activate should send an event with a valid flag-key for feature-test.

File: Sources/Data Model/DispatchEvents/BatchEvent.swift [hunk 10]
```diff
@@ -104,7 +120,7 @@ struct DispatchEvent: Codable, Equatable {
     var tags: [String: AttributeValue]?
     var revenue: AttributeValue?
     var value: AttributeValue?
-
+
```
Comment (msohailhussain) [inline Sources/Data Model/DispatchEvents/BatchEvent.swift:123]: remove these extra spaces.

File: Sources/Data Model/DispatchEvents/BatchEvent.swift [hunk 11]
```diff
@@ -124,7 +140,7 @@ struct DispatchEvent: Codable, Equatable {
          revenue: AttributeValue? = nil) {
         
         // TODO: add validation and throw here for invalid value (int, double) and revenue (int) types
-
+
```
Comment (msohailhussain) [inline Sources/Data Model/DispatchEvents/BatchEvent.swift:143]: remove these extra spaces.

File: Sources/Implementation/DefaultDecisionService.swift [hunk 12]
```diff
@@ -41,7 +41,7 @@ class DefaultDecisionService: OPTDecisionService {
         
         // ---- check if the user is forced into a variation ----
         if let variationId = config.getForcedVariation(experimentKey: experiment.key, userId: userId)?.id,
-            let variation = experiment.getVariation(id: variationId) {
+           let variation = experiment.getVariation(id: variationId) {
```
Comment (msohailhussain) [inline Sources/Implementation/DefaultDecisionService.swift:44]: please fix indentation.

**Review by msohailhussain — CHANGES_REQUESTED** [review]
look good, but please fix indentation issues, empty spaces and i saw some other spacing issues.

File: Sources/Data Model/DispatchEvents/BatchEvent.swift (hunk 10 above)
Comment (yasirfolio3) [reply Sources/Data Model/DispatchEvents/BatchEvent.swift:123]: this spacing is being added automatically by the new xcode, it follows the same spacing as the previous line.

**Review by msohailhussain — DISMISSED** [review]
lgtm. @jaeopt please check indentation rules are changed in xcode 12.

File: Sources/Implementation/DefaultDecisionService.swift [hunk 13]
```diff
@@ -147,6 +147,7 @@ class DefaultDecisionService: OPTDecisionService {
         }
         
         return nil
+
```
Comment (jaeopt) [inline Sources/Implementation/DefaultDecisionService.swift]: Remove this empty line

File: Sources/Optimizely/OptimizelyClient.swift [hunk 14]
```diff
@@ -259,7 +259,7 @@ open class OptimizelyClient: NSObject {
                             variation: variation,
                             userId: userId,
                             attributes: attributes,
-                            flagKey: experimentKey,
+                            flagKey: "",
                             ruleType: "experiment")
```
Comment (jaeopt) [inline Sources/Optimizely/OptimizelyClient.swift:263]: Do we have this new ruleType "experiment"? If so, can we add it to DecisionSource enum?

File: Tests/OptimizelyTests-Common/DecisionListenerTests.swift [hunk 15]
```diff
@@ -778,6 +780,19 @@ class DecisionListenerTests: XCTestCase {
...
+        let snapshot = (visitor["snapshots"] as! Array<Dictionary<String, Any>>)[0]
+        let decision = (snapshot["decisions"]  as! Array<Dictionary<String, Any>>)[0]
+        
+        let metaData = decision["metadata"] as! Dictionary<String, Any>
+        XCTAssertEqual(metaData["rule_type"] as! String, Constants.DecisionSource.rollout.rawValue)
+        XCTAssertEqual(metaData["rule_key"] as! String, "")
+        XCTAssertEqual(metaData["flag_key"] as! String, "feature_1")
+        XCTAssertEqual(metaData["variation_key"] as! String, "")
+        
+        self.optimizely.config!.project!.sendFlagDecisions = nil
```
Comment (jaeopt) [inline Tests/OptimizelyTests-Common/DecisionListenerTests.swift:795]: This test looks good, but looks like in a wrong place, not related to Notification. Can we move this part to other like BathEventTests?

**Review by jaeopt — CHANGES_REQUESTED** [review]
A couple of small changes suggested. Other than that, it looks good!

File: Tests/OptimizelyTests-Common/BatchEventBuilderTests_Events.swift [hunk 16]
```diff
@@ -249,17 +250,119 @@ class BatchEventBuilderTests_Events: XCTestCase {
...
+                                            eventDispatcher: eventDispatcher2)
+        try! fakeOptimizelyManager.start(datafile: datafile)
+        
+        let exp = expectation(description: "Wait for event to dispatch")
+        fakeOptimizelyManager.config!.project!.sendFlagDecisions = true
+        fakeOptimizelyManager.setDecisionServiceData(experiment: nil, variation: nil, source: "")
+        _ = fakeOptimizelyManager.isFeatureEnabled(featureKey: featureKey, userId: userId)
+        
+        let result = XCTWaiter.wait(for: [exp], timeout: 0.1)
+         if result == XCTWaiter.Result.timedOut {
```
Comment (jaeopt) [inline Tests/OptimizelyTests-Common/BatchEventBuilderTests_Events.swift:269]: Fix indentation in all three tests

**Review by jaeopt — APPROVED** [review]
LGTM - nit: fix indent

File: Tests/OptimizelyTests-APIs/OptimizelyClientTests_Others.swift [hunk 17]
```diff
@@ -290,7 +290,7 @@ class OptimizelyClientTests_Others: XCTestCase {
         // set invalid (infinity) to attribute values, which will cause JSONEncoder.encode exception
         let attributes = ["testvar": Double.infinity]
         
-        optimizely.sendImpressionEvent(experiment: experiment, variation: variation, userId: kUserId, attributes: attributes)
+        optimizely.sendImpressionEvent(experiment: experiment, variation: variation, userId: kUserId, attributes: attributes, flagKey: experiment.key, ruleType: Constants.DecisionSource.rollout.rawValue)
```
Comment (pawels-optimizely) [inline Tests/OptimizelyTests-APIs/OptimizelyClientTests_Others.swift:293]: don't assign experiment.key to flagKey even in tests

**Review by pawels-optimizely — DISMISSED** [review]
looks good, one nit comment, I hope it ran through FSC tests.

**Comment by yasirfolio3** [comment]
> looks good, one nit comment, I hope it ran through FSC tests.

suggested changes made, Have also verified it with fsc flag-decision tests.

**Review by msohailhussain — APPROVED** [review]
lgtm

**Review by pawels-optimizely — APPROVED** [review]
lgtm

**Review by jaeopt — APPROVED** [review]
LGTM
